dotnet/src/Experimental/Process.Runtime.Dapr/Serialization/KernelProcessEventSerializer.cs

```diff
@@ -0,0 +1,38 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+using System.Collections.Generic;
+using System.Text.Json;
+
+namespace Microsoft.SemanticKernel.Process.Serialization;
+
+/// <summary>
+/// Serializer for <see cref="KernelProcessEvent"/> objects.
+/// </summary>
+/// <remarks>
+/// Includes type info for <see cref="KernelProcessEvent.Data"/>.
+/// </remarks>
+internal static class KernelProcessEventSerializer
+{
+    /// <summary>
+    /// Serialize <see cref="KernelProcessEvent"/> to JSON with type information.
+    /// </summary>
+    public static string ToJson(this KernelProcessEvent processEvent)
+    {
+        EventContainer<KernelProcessEvent> containedEvents = new(TypeInfo.GetAssemblyQualifiedType(processEvent.Data), processEvent);
+        return JsonSerializer.Serialize(containedEvents);
+    }
+
+    /// <summary>
+    /// Deserialize a list of JSON events into a list of <see cref="KernelProcessEvent"/> objects.
+    /// </summary>
+    /// <exception cref="KernelException"></exception>
+    public static IEnumerable<KernelProcessEvent> ToKernelProcessEvents(this IEnumerable<string> jsonEvents)
+    {
+        foreach (string json in jsonEvents)
+        {
+            EventContainer<KernelProcessEvent> eventContainer =
+                JsonSerializer.Deserialize<EventContainer<KernelProcessEvent>>(json) ??
+                throw new KernelException($"Unable to deserialize {nameof(KernelProcessEvent)} queue.");
```

Comment:
just wondering what would happen here if one of the events was unable to deserialize and the others were ok.
The good apples would not get delivered right - only the ones before the bad apple?

Reply:
In general should we "dump"/skip bad messages and move on or throw an exception that would freeze everything?

Reply:
Also thinking what could potentially made this happen from the dapr side

Reply:
discussed offline

Reply:
Good call-out, from our offline discussion:

> Yeah, that's a good question.  I don't think we want to ignore the "bad apples" and unclear the value in delivering "good apples" in the face of framework error...but yeah, its kind'of right in the middle...which admittedly has ambiguous impact.
I suppose the two mitigators are:
> 1. We manage this contract entirely.  If serialization fails, something has gone terribly wrong with our internal consistency.  I'd honestly not expect this arc to be hit...but the serializer returns null...so it has to be dealt with in code.
> 2. In the call-sites, I'm pushing the entire enumerated result into array (`ToArray`)...so one bad apple should fail the larger operation prior to taking action. (where `foreach` could provide indeterminate results).

I've adjusted the serializer to process the entire enumeration (moved `ToArray` from the caller to the serializer)...this ensure consistency at all call points.  I think this will ensure consistency with the resulting behavior should the data-contract serialization had failed.  Also, if the developer provides state that is not serializable...I believe we'd experience failure prior to this (on `ToJson`).

I agree with your offline point that Dapr serialization will need some notes / documentation.